### DISCUSSION TITLE
ContactListener firing ContactAdded over and over on HeightField

### DISCUSSION THREAD

**Discussion by DennisSmolek** [jrouwe/JoltPhysics.js#127] [Q&A]
I'm currently in the process of getting the heightfield system to work on my project.

I have the visual of the heightfield correct but the Jolt version isn't lining up, (scale/size things)

While working through that I thought it would be useful to detect if a sphere body intersects the heightfield.

However I'm getting a really weird behavior. As the sphere rolls on the heightfield it fires the `onContactAdded` and `OnContactEnded` events over and over.
![image](https://github.com/jrouwe/JoltPhysics.js/assets/1397052/b1a175bc-b6d5-4dae-87fd-797a4bdd1cf4)
(first number is handle of body1, second is handle of body2)

Sometimes it fires multiple in a row of both start and ending events. 

Now it could be something related to my still broken implementation which I'll have to work on without contact events, however it's definitely something to do with heightfields. 

I tested the same scene on an angled plane and when the spheres roll along a plane they just fire `OnContactPersisted` as expected.

I remember reading something in the API that a heightfield can detect where it's being contacted or something, so it makes me wonder if a contact event is firing for every block of the heightfield and not the heightfield as a whole...
That would match up to the behavior I'm seeing.

With my implementation not being 100% I'm not sure this is a bug, so if anyone has any insights I'd greatly appreciate it!

**Comment by jrouwe** [comment]
The contact listener always gives you a pair of (body id, sub shape id) for both bodies. The sub shape ID is stored in `ContactManifold::mSubShapeID1/2`. Each triangle (that is generated on the fly) of the height field has its own sub shape id, so as your sphere rolls over the height field it will tell you 'contact added between heightfield H and sub shape ID X' and the next frame you may receive 'contact added between heightfield H and sub shape ID Y' followed by 'contact lost between heightfield H and sub shape ID X' (in that order) so if you're interested in knowing if two objects have any contacts remaining then do something like:

```
OnContactAdded(body1, body2)
lock mutex because this gets called from multiple threads
if (++num_contacts[pair(body1, body2)] == 0)
  first contact between body1 and body2 was added

OnContactRemoved(body1, body2)
lock mutex because this gets called from multiple threads
if (--num_contacts[pair(body1, body2)] == 0)
  last contact between body1 and body 2 was removed
```

where num_contacts could be a hash map of pair<BodyID, BodyID> -> int.

**Reply by jrouwe** [reply]
(Oh and my answer is very much geared towards C++, in JS you currently don't have multi threading so no need to lock a mutex)

**Reply by DennisSmolek** [reply]
Very Interesting!

So each block in the Heightfield is considered a SubShape and the `contactListener` isn't listening if a `Jolt.Body` contacts but if the `.Jolt.SubShapeID` Does.

So this means for compound shapes too, that a body could have multiple contacts with it as well. 

Two follow up questions:

1. Before writing my own similar to you suggested, Is there an existing method to test if two `Body`'s are currently in contact? Something like: `Body.CheckContacting(BodyID2)`
I would use a function like this for something like checking if a character is on the ground `onGround(heightfeildId)` or to validate if two bodies are in contact etc. Because the listener fires for more than the just body it can't be used without verifying the number of contacts.

2. With the added complexity of the heightfield and it's subshapes, would it be more performant to use a TriangleMesh for certain terrain like features? I'm guessing the mesh would only have a single subshape to test against.

**Reply by jrouwe** [reply]
> So this means for compound shapes too, that a body could have multiple contacts with it as well.

Correct.

> Before writing my own similar to you suggested, Is there an existing method to test if two Body's are currently in contact?

Yes there is, I forgot about it. It's on PhysicsSystem::WereBodiesInContact and I just submitted a change that exposes it.

> I'm guessing the mesh would only have a single subshape to test against.

Every triangle in a mesh also has its own subshape ID. This allows you to do things like modifying the friction based on which part of the mesh you're hitting.

**Reply by DennisSmolek** [reply]
Very cool,  I just found the Docs for `wereBodiesInContact` so I'll do a build and give that a shot. 


> so as your sphere rolls over the height field it will tell you 'contact added between heightfield H and sub shape ID X' and the next frame you may receive 'contact added between heightfield H and sub shape ID Y' followed by 'contact lost between heightfield H and sub shape ID X' (in that order)

Unfortunately my trial at an implementation has the events firing in a different order:
```
1711008019466 ': NEW Contact Added' 16777216 16777217 1
1711008019499 ': FINAL Contact Removed' 16777216 16777217 0
1711008019665 ': NEW Contact Added' 16777216 16777217 1
1711008019682 ': FINAL Contact Removed' 16777216 16777217 0
1711008019783 ': NEW Contact Added' 16777216 16777217 1
1711008019832 ': FINAL Contact Removed' 16777216 16777217 0
1711008020399 ': NEW Contact Added' 16777216 16777217 1
1711008020433 ': FINAL Contact Removed' 16777216 16777217 0
1711008021133 ': NEW Contact Added' 16777216 16777217 1
1711008021167 ': FINAL Contact Removed' 16777216 16777217 0
1711008021784 ': NEW Contact Added' 16777216 16777217 1
1711008021851 ': FINAL Contact Removed' 16777216 16777217 0
1711008022150 ': NEW Contact Added' 16777216 16777217 1
1711008022201 ': FINAL Contact Removed' 16777216 16777217 0
```

Where the large numbers are the bodyId's and the last number is the number of active contacts. 
(I'm skipping new contacts added/removed between the finals in this log)

I did much more detailed logging but it says essentially the same thing, on occasion as the ball is rolling it occupies a single block, and then as it goes to the next block, it fires that it's leaving THEN it fires that it'ls joining a new subshape. 

I tried logging out the subshapes from the manifold but `SubShapeID.GetValue()` seems to only return `-1`

I'm going to shelve this for now to work on other parts but at the moment I don't have a reliable way to detect when two bodies start/stop contacting (at least those with subShapes).
I'm considering adding a timestamp/threshold which would probably work...

**Reply by jrouwe** [reply]
This is not what I'm seeing and also technically quite impossible, the job graph looks like this:

![image](https://github.com/jrouwe/JoltPhysics.js/assets/1621693/05252274-1223-4d78-838c-3d77bc7cd9b6)

As you can see the contact added callbacks are generated well before the contact remove callbacks.

If I update the HeightFieldShapeTest in the C++ samples app to spawn a sphere instead of a box and go to `Physics Settings` and do `Install Contact Listener` then it traces remove always after add, e.g.:

```
Draw Contact State
Validate 1 and 0 result 0
Contact persisted 0 (ffffbf7d) and 1 (ffffffff)
Draw Contact State
Validate 1 and 0 result 0
Contact added 0 (ffffbf7c) and 1 (ffffffff)
Contact removed 0 (ffffbf7d) and 1 (ffffffff)
Draw Contact State
```

> I tried logging out the subshapes from the manifold but SubShapeID.GetValue() seems to only return -1

The sub shape for the sphere should always be -1, the one for the heightfield should not (as you can see in the number between brackets above)

Perhaps your sphere has a bit of restitution and bounces a bit before going to the next triangle?
